core/src/Cabana_ParticleGridCommunication.hpp

```diff
@@ -0,0 +1,790 @@
+                 positions( p, Cajita::Dim::I ) > local_high[Cajita::Dim::I] ||
+                 positions( p, Cajita::Dim::J ) < local_low[Cajita::Dim::J] ||
+                 positions( p, Cajita::Dim::J ) > local_high[Cajita::Dim::J] ||
+                 positions( p, Cajita::Dim::K ) < local_low[Cajita::Dim::K] ||
+                 positions( p, Cajita::Dim::K ) > local_high[Cajita::Dim::K] )
+                result += 1;
+        },
+        comm_count );
+
+    MPI_Allreduce( MPI_IN_PLACE, &comm_count, 1, MPI_INT, MPI_SUM,
+                   local_grid.globalGrid().comm() );
+
+    return comm_count;
+}
+
+namespace Impl
+{
+//---------------------------------------------------------------------------//
+// Locate the particles in the local grid and get their destination rank.
+// Particles are assumed to only migrate to a location in the 26 neighbor halo
+// or stay on this rank. If the particle crosses a global periodic boundary,
+// wrap it's coordinates back into the domain.
+template <class LocalGridType, class PositionSliceType, class NeighborRankView,
+          class DestinationRankView>
+void getMigrateDestinations( const LocalGridType& local_grid,
+                             const NeighborRankView& neighbor_ranks,
+                             DestinationRankView& destinations,
+                             const PositionSliceType& positions )
+{
+    using execution_space = typename PositionSliceType::execution_space;
+
+    const auto& local_mesh =
+        Cajita::createLocalMesh<Kokkos::HostSpace>( local_grid );
+
+    // Check within the local domain.
+    const Kokkos::Array<double, 3> local_low = {
+        local_mesh.lowCorner( Cajita::Own(), Cajita::Dim::I ),
+        local_mesh.lowCorner( Cajita::Own(), Cajita::Dim::J ),
+        local_mesh.lowCorner( Cajita::Own(), Cajita::Dim::K ) };
+    const Kokkos::Array<double, 3> local_high = {
+        local_mesh.highCorner( Cajita::Own(), Cajita::Dim::I ),
+        local_mesh.highCorner( Cajita::Own(), Cajita::Dim::J ),
+        local_mesh.highCorner( Cajita::Own(), Cajita::Dim::K ) };
+
+    Kokkos::parallel_for(
+        "get_migrate_destinations",
+        Kokkos::RangePolicy<execution_space>( 0, positions.size() ),
+        KOKKOS_LAMBDA( const int p ) {
+            // Compute the logical index of the neighbor we are sending to.
+            int nid[3] = { 1, 1, 1 };
+            for ( int d = 0; d < 3; ++d )
+            {
+                if ( positions( p, d ) < local_low[d] )
+                    nid[d] = 0;
+                else if ( positions( p, d ) > local_high[d] )
+                    nid[d] = 2;
+            }
+
+            // Compute the destination MPI rank.
+            destinations( p ) = neighbor_ranks(
+                nid[Cajita::Dim::I] +
+                3 * ( nid[Cajita::Dim::J] + 3 * nid[Cajita::Dim::K] ) );
+        } );
+    // TODO: fuse kernels
+    periodicWrap( local_grid, positions );
```

Comment:
Why not fuse them in this PR? They are already fused in the Picasso implementation.

Reply:
Forgot about that - done